### PR TITLE
F252: token request clean up 

### PR DESCRIPTION
**Pull request by KElkhiyaoui**
- add nil checks

### CODE REVIEW

File: token/driver/request.go [hunk 1]
```diff
@@ -93,84 +94,123 @@ type TokenRequestMetadata struct {
 	Application map[string][]byte
 }
 
-func (m *TokenRequestMetadata) TokenInfos() [][]byte {
+func (m *TokenRequestMetadata) TokenInfos() ([][]byte, error) {
 	var res [][]byte
-	for _, issue := range m.Issues {
+	for j, issue := range m.Issues {
+		if &issue == nil {
+			return nil, errors.Errorf("cannot get token info: nil issue index [%d]", j)
```
Comment (HagarMeir) [inline token/driver/request.go:101]: nil issue **at** index

File: token/driver/request.go [hunk 2]
```diff
@@ -93,84 +94,123 @@ type TokenRequestMetadata struct {
...
 	var res [][]byte
-	for _, issue := range m.Issues {
-		for _, r := range issue.Receivers {
+	for j, issue := range m.Issues {
+		if &issue == nil {
+			return nil, errors.Errorf("cannot get recipients: nil issue index [%d]", j)
+		}
+		for i, r := range issue.Receivers {
+			if r == nil {
+				return nil, errors.Errorf("cannot serialize [%dth] recipient in issue at index [%d]: nil recipient", i, j)
```
Comment (HagarMeir) [inline token/driver/request.go:146]: nil **receiver**

File: token/metadata.go [hunk 3]
```diff
@@ -30,7 +29,16 @@ type Metadata struct {
 
 // GetToken unmarshals the given bytes to extract the token and its issuer (if any).
 func (m *Metadata) GetToken(raw []byte) (*token.Token, view.Identity, []byte, error) {
-	tokenInfoRaw := m.TokenRequestMetadata.GetTokenInfo(raw)
+	if m.TokenRequestMetadata == nil {
+		return nil, nil, nil, errors.New("failed to get token: nil metadata")
+	}
+	if m.TMS == nil {
+		return nil, nil, nil, errors.New("failed to get token: initialized the token service")
```
Comment (HagarMeir) [inline token/metadata.go:36]: nil token service

File: token/metadata.go [hunk 4]
```diff
@@ -30,7 +29,16 @@ type Metadata struct {
...
+	}
+	tokenInfoRaw, err := m.TokenRequestMetadata.GetTokenInfo(raw)
+	if err != nil {
+		return nil, nil, nil, errors.Wrap(err, "failed to get token")
```
Comment (HagarMeir) [inline token/metadata.go:40]: failed to get token info

File: token/metadata.go [hunk 5]
```diff
@@ -61,17 +75,27 @@ func (m *Metadata) SpentTokenID() []*token.ID {
 // - The senders are included if and only if there is at least one output whose owner has the given enrollment IDs.
 // Application metadata is always included
 func (m *Metadata) FilterBy(eIDs ...string) (*Metadata, error) {
+	if m.TMS == nil || m.TokenRequestMetadata == nil {
+		return nil, errors.New("can't filter metadata by eID: invalid metadata")
```
Comment (HagarMeir) [inline token/metadata.go:79]: or nil token service

File: token/metadata.go [hunk 6]
```diff
@@ -61,17 +75,27 @@ func (m *Metadata) SpentTokenID() []*token.ID {
...
 	res := &Metadata{
 		TMS:                  m.TMS,
 		TokenRequestMetadata: &driver.TokenRequestMetadata{},
 	}
 
 	// filter issues
-	for _, issue := range m.TokenRequestMetadata.Issues {
+	for j, issue := range m.TokenRequestMetadata.Issues {
+		if &issue == nil {
+			return nil, errors.Errorf("can't filer metadata by eID: issue at index [%d] is nil", j)
```
Comment (HagarMeir) [inline token/metadata.go:89]: **filter**

File: token/metadata.go [hunk 7]
```diff
@@ -102,9 +126,11 @@ func (m *Metadata) FilterBy(eIDs ...string) (*Metadata, error) {
 	}
 
 	// filter transfers
-	for _, transfer := range m.TokenRequestMetadata.Transfers {
+	for j, transfer := range m.TokenRequestMetadata.Transfers {
 		transferRes := driver.TransferMetadata{}
-
+		if &transfer == nil {
+			return nil, errors.Errorf("can't filer metadata by eID: transfer at index [%d] is nil", j)
```
Comment (HagarMeir) [inline token/metadata.go:132]: **filter**

File: token/metadata.go [hunk 8]
```diff
@@ -119,6 +145,11 @@ func (m *Metadata) FilterBy(eIDs ...string) (*Metadata, error) {
 			var ReceiverIsSender bool
 			var ReceiverAuditInfos []byte
 
+			if len(transfer.Outputs) != len(transfer.ReceiverAuditInfos) || len(transfer.TokenInfo) != len(transfer.ReceiverAuditInfos) ||
+				len(transfer.Receivers) != len(transfer.ReceiverAuditInfos) || len(transfer.ReceiverIsSender) != len(transfer.ReceiverAuditInfos) {
+				return nil, errors.Errorf("can't filer metadata by eID: transfer at index [%d] is invalid", j)
```
Comment (HagarMeir) [inline token/metadata.go:150]: **filter**

File: token/metadata.go [hunk 9]
```diff
@@ -137,10 +168,14 @@ func (m *Metadata) FilterBy(eIDs ...string) (*Metadata, error) {
 			transferRes.ReceiverAuditInfos = append(transferRes.ReceiverAuditInfos, ReceiverAuditInfos)
 			transferRes.TokenInfo = append(transferRes.TokenInfo, TokenInfo)
 		}
+
 		// if skip = true, it means that this transfer does not contain any output for the given enrollment IDs.
 		// Therefore, no metadata should be given to the passed enrollment IDs.
 		// if skip = false, it means that this transfer contains at least one output for the given enrollment IDs.
 		// Append the senders to the transfer metadata.
+		if len(transfer.Senders) != len(transfer.SenderAuditInfos) {
+			return nil, errors.Errorf("can't filer metadata by eID: transfer at index [%d] is invalid", j)
```
Comment (HagarMeir) [inline token/metadata.go:177]: **filter**

File: token/metadata.go [hunk 10]
```diff
@@ -185,7 +226,11 @@ type IssueMetadata struct {
 }
 
 // Match returns true if the given action matches this metadata
+// todo is match a good name for this?
```
Comment (HagarMeir) [inline token/metadata.go:229]: please address

File: token/services/ttx/endorse.go [hunk 11]
```diff
@@ -207,7 +210,10 @@ func (c *collectEndorsementsView) requestSignaturesOnIssues(context view.Context
 }
 
 func (c *collectEndorsementsView) requestSignaturesOnTransfers(context view.Context) ([]view.Identity, error) {
-	transfers := c.tx.TokenRequest.Transfers()
+	transfers, err := c.tx.TokenRequest.Transfers()
+	if err != nil {
+		return nil, errors.Wrap(err, "request signatures on transfers failed")
```
Comment (HagarMeir) [inline token/services/ttx/endorse.go:215]: failed to get transfers

File: token/services/ttx/endorse.go [hunk 12]
```diff
@@ -112,7 +112,10 @@ func (c *collectEndorsementsView) Call(context view.Context) (interface{}, error
 
 func (c *collectEndorsementsView) requestSignaturesOnIssues(context view.Context) ([]view.Identity, error) {
 
-	issues := c.tx.TokenRequest.Issues()
+	issues, err := c.tx.TokenRequest.Issues()
+	if err != nil {
+		return nil, errors.Wrap(err, "requestSignaturesOnIssues failed")
```
Comment (HagarMeir) [inline token/services/ttx/endorse.go:117]: failed to get issues

File: token/request.go [hunk 13]
```diff
@@ -782,6 +896,7 @@ func (t *Request) AuditInputs() (*InputStream, error) {
 
 	// load the tokens corresponding to the input token ids
 	ids := inputs.IDs()
+	// todo check token service
```
Comment (HagarMeir) [inline token/request.go:899]: please address

File: token/driver/request.go [hunk 14]
```diff
@@ -93,84 +94,123 @@ type TokenRequestMetadata struct {
...
+	return res, nil
 }
 
-func (m *TokenRequestMetadata) Senders() [][]byte {
+func (m *TokenRequestMetadata) Senders() ([][]byte, error) {
 	var res [][]byte
-	for _, transfer := range m.Transfers {
-		for _, s := range transfer.Senders {
+	for j, transfer := range m.Transfers {
+		if &transfer == nil {
```
Comment (adecaro) [inline token/driver/request.go:168]: as above

File: token/driver/request.go [hunk 15]
```diff
@@ -93,84 +94,123 @@ type TokenRequestMetadata struct {
...
-	return res
+	return res, nil
 }
 
-func (m *TokenRequestMetadata) Inputs() []*token2.ID {
+func (m *TokenRequestMetadata) Inputs() ([]*token2.ID, error) {
 	var res []*token2.ID
-	for _, transfer := range m.Transfers {
+	for j, transfer := range m.Transfers {
+		if &transfer == nil {
```
Comment (adecaro) [inline token/driver/request.go:195]: as above

File: token/driver/request.go [hunk 16]
```diff
@@ -93,84 +94,123 @@ type TokenRequestMetadata struct {
...
 func (m *TokenRequestMetadata) Bytes() ([]byte, error) {
 	meta, err := MarshalMeta(m.Application)
 	if err != nil {
-		return nil, err
+		return nil, errors.New("cannot marshal token request metadata: failed to marshal application metadata")
 	}
 
 	transfers := make([]TransferMetadataSer, len(m.Transfers))
 	for i, transfer := range m.Transfers {
+		if &transfer == nil {
```
Comment (adecaro) [inline token/driver/request.go:211]: as above

File: token/driver/request.go [hunk 17]
```diff
@@ -199,14 +239,20 @@ func (m *TokenRequestMetadata) FromBytes(raw []byte) error {
...
 
 	m.Issues = ser.Issues
 	m.Transfers = make([]TransferMetadata, len(ser.Transfers))
 	for i, transfer := range ser.Transfers {
+		if &transfer == nil {
+			return errors.Errorf("failed to unmarshal token request metadata: nil transfer at index [%d]", i)
+		}
 		TokenIDs := make([]*token2.ID, len(transfer.TokenIDs))
 		for j, tokenID := range transfer.TokenIDs {
+			if &tokenID == nil {
```
Comment (adecaro) [inline token/driver/request.go:253]: as bove

File: token/services/network/fabric/processor.go [hunk 18]
```diff
@@ -149,7 +149,11 @@ func (r *RWSetProcessor) tokenRequest(req fabric.Request, tx fabric.ProcessTrans
 
 	if tms.PublicParametersManager().GraphHiding() {
 		// Delete inputs
-		for _, id := range metadata.SpentTokenID() {
+		ids, err := metadata.SpentTokenID()
```
Comment (adecaro) [inline token/services/network/fabric/processor.go:152]: checking SpentTokenID, I think we don't need this error here

File: token/services/ttx/endorse.go [hunk 19]
```diff
@@ -739,7 +745,11 @@ func (s *endorseView) Call(context view.Context) (interface{}, error) {
 
 func (s *endorseView) requestsToBeSigned() ([]*token.Transfer, error) {
 	var res []*token.Transfer
-	for _, transfer := range s.tx.TokenRequest.Transfers() {
+	transfers, err := s.tx.TokenRequest.Transfers()
```
Comment (adecaro) [inline token/services/ttx/endorse.go:748]: lokking at the Transfers, we don't need this error here

File: token/metadata.go [hunk 20]
```diff
@@ -43,12 +51,18 @@ func (m *Metadata) GetToken(raw []byte) (*token.Token, view.Identity, []byte, er
...
 // SpentTokenID returns the token IDs of the tokens that were spent by the Token Request this metadata is associated with.
-func (m *Metadata) SpentTokenID() []*token.ID {
+func (m *Metadata) SpentTokenID() ([]*token.ID, error) {
 	var res []*token.ID
-	for _, transfer := range m.TokenRequestMetadata.Transfers {
+	if m.TokenRequestMetadata == nil {
+		return nil, errors.New("can't get spent token ID: nil metadata")
+	}
+	for i, transfer := range m.TokenRequestMetadata.Transfers {
+		if &transfer == nil {
```
Comment (adecaro) [inline token/metadata.go:60]: no need for this check

**Review by adecaro — CHANGES_REQUESTED** [review]

File: token/driver/request.go (hunk 2 above)
Comment (KElkhiyaoui) [reply token/driver/request.go:146]: OK

File: token/metadata.go (hunk 5 above)
Comment (KElkhiyaoui) [reply token/metadata.go:79]: OK

File: token/services/ttx/endorse.go (hunk 12 above)
Comment (KElkhiyaoui) [reply token/services/ttx/endorse.go:117]: OK

File: token/request.go (hunk 13 above)
Comment (KElkhiyaoui) [reply token/request.go:899]: OK

File: token/driver/request.go [hunk 21]
```diff
@@ -93,84 +94,123 @@ type TokenRequestMetadata struct {
...
-	for _, issue := range m.Issues {
+	for j, issue := range m.Issues {
+		if &issue == nil {
+			return nil, errors.Errorf("cannot get token info: nil issue index [%d]", j)
+		}
 		res = append(res, issue.TokenInfo...)
 	}
-	for _, transfer := range m.Transfers {
+	for j, transfer := range m.Transfers {
+		if &transfer == nil {
```
[the remark this answers is not in the document]
Comment (KElkhiyaoui) [reply token/driver/request.go:106]: OK

File: token/driver/request.go (hunk 14 above)
Comment (KElkhiyaoui) [reply token/driver/request.go:168]: OK

File: token/driver/request.go [hunk 22]
```diff
@@ -93,84 +94,123 @@ type TokenRequestMetadata struct {
...
-	return res
+	return res, nil
 }
 
-func (m *TokenRequestMetadata) Issuers() [][]byte {
+func (m *TokenRequestMetadata) Issuers() ([][]byte, error) {
 	var res [][]byte
-	for _, issue := range m.Issues {
+	for j, issue := range m.Issues {
+		if &issue == nil {
```
[the remark this answers is not in the document]
Comment (KElkhiyaoui) [reply token/driver/request.go:184]: OK

File: token/driver/request.go (hunk 15 above)
Comment (KElkhiyaoui) [reply token/driver/request.go:195]: OK

File: token/driver/request.go (hunk 16 above)
Comment (KElkhiyaoui) [reply token/driver/request.go:211]: OK

File: token/driver/request.go [hunk 23]
```diff
@@ -199,14 +239,20 @@ func (m *TokenRequestMetadata) FromBytes(raw []byte) error {
...
 	_, err := asn1.Unmarshal(raw, ser)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "failed to unmarshal token request metadata")
 	}
 
 	m.Issues = ser.Issues
 	m.Transfers = make([]TransferMetadata, len(ser.Transfers))
 	for i, transfer := range ser.Transfers {
+		if &transfer == nil {
```
[the remark this answers is not in the document]
Comment (KElkhiyaoui) [reply token/driver/request.go:248]: OK

File: token/driver/request.go [hunk 24]
```diff
@@ -93,84 +94,123 @@ type TokenRequestMetadata struct {
...
+func (m *TokenRequestMetadata) Senders() ([][]byte, error) {
 	var res [][]byte
-	for _, transfer := range m.Transfers {
-		for _, s := range transfer.Senders {
+	for j, transfer := range m.Transfers {
+		if &transfer == nil {
+			return nil, errors.Errorf("cannot get senders: nil transfer index [%d]", j)
+		}
+		for i, s := range transfer.Senders {
+			if s == nil {
```
[the remark this answers is not in the document]
Comment (KElkhiyaoui) [reply token/driver/request.go:172]: OK

File: token/driver/request.go [hunk 25]
```diff
@@ -93,84 +94,123 @@ type TokenRequestMetadata struct {
...
 		}
 	}
-	for _, transfer := range m.Transfers {
-		for _, r := range transfer.Receivers {
+	for j, transfer := range m.Transfers {
+		if &transfer == nil {
+			return nil, errors.Errorf("cannot get recipients: nil transfer index [%d]", j)
+		}
+		for i, r := range transfer.Receivers {
+			if r == nil {
```
[the remark this answers is not in the document]
Comment (KElkhiyaoui) [reply token/driver/request.go:156]: noted

File: token/services/network/orion/processor.go [hunk 26]
```diff
@@ -115,7 +115,11 @@ func (r *RWSetProcessor) tokenRequest(req orion.Request, tx orion.ProcessTransac
 
 	if tms.PublicParametersManager().GraphHiding() {
 		// Delete inputs
-		for _, id := range metadata.SpentTokenID() {
+		ids, err := metadata.SpentTokenID()
+		if err != nil {
+			return err
```
Comment (adecaro) [inline token/services/network/orion/processor.go:120]: better error message?

File: token/services/network/fabric/processor.go [hunk 27]
```diff
@@ -149,7 +149,11 @@ func (r *RWSetProcessor) tokenRequest(req fabric.Request, tx fabric.ProcessTrans
...
+		if err != nil {
+			return err
```
Comment (adecaro) [inline token/services/network/fabric/processor.go:154]: better error message?

File: token/metadata.go [hunk 28]
```diff
@@ -61,17 +69,24 @@ func (m *Metadata) SpentTokenID() []*token.ID {
...
 	}
 
 	// filter issues
-	for _, issue := range m.TokenRequestMetadata.Issues {
+	for j, issue := range m.TokenRequestMetadata.Issues {
 		issueRes := driver.IssueMetadata{
 			Issuer: issue.Issuer,
 		}
 
+		if len(issue.ReceiversAuditInfos) != len(issue.Outputs) || len(issue.ReceiversAuditInfos) != len(issue.TokenInfo) ||
```
Comment (adecaro) [inline token/metadata.go:86]: Do we need this check here? Should we have a function that validates the metadata instead?

File: token/request.go [hunk 29]
```diff
@@ -481,8 +502,23 @@ func (t *Request) Inputs() (*InputStream, error) {
 }
 
 func (t *Request) inputs(failOnMissing bool) (*InputStream, error) {
+	if t.TokenService == nil || t.TokenService.tms == nil {
```
Comment (adecaro) [inline token/request.go:505]: if these checks fail, we have a big problem no? Either something is not properly initialized, therefore a bug must be fixed, or the system is not responding as expected. Does it make sense to have a function to validate the metadata and that can be used once before any other operation is performed?

File: token/request.go [hunk 30]
```diff
@@ -873,6 +1003,14 @@ func (t *Request) countOutputs() (int, error) {
 }
 
 func (t *Request) parseInputIDs(inputs []*token.ID) ([]*token.ID, token.Quantity, string, error) {
+	if t.TokenService == nil || t.TokenService.Vault() == nil || t.TokenService.Vault().NewQueryEngine() == nil {
```
Comment (adecaro) [inline token/request.go:1006]: shall avoid this?

**Review by adecaro — CHANGES_REQUESTED** [review]

File: token/services/network/fabric/processor.go (hunk 27 above)
Comment (KElkhiyaoui) [reply token/services/network/fabric/processor.go:154]: OK

File: token/services/network/orion/processor.go (hunk 26 above)
Comment (KElkhiyaoui) [reply token/services/network/orion/processor.go:120]: OK

File: token/metadata.go [hunk 31]
```diff
@@ -61,17 +69,24 @@ func (m *Metadata) SpentTokenID() []*token.ID {
 // - The senders are included if and only if there is at least one output whose owner has the given enrollment IDs.
 // Application metadata is always included
 func (m *Metadata) FilterBy(eIDs ...string) (*Metadata, error) {
+	if m.TMS == nil || m.TokenRequestMetadata == nil {
```
[the remark this answers is not in the document]
Comment (KElkhiyaoui) [reply token/metadata.go:72]: I will remove the checks, as this will never happen, looking at the method calls.

File: token/request.go (hunk 29 above)
Comment (KElkhiyaoui) [reply token/request.go:505]: I agree - from the code, this will never happen.

File: token/request.go (hunk 30 above)
Comment (KElkhiyaoui) [reply token/request.go:1006]: I agree - I checked the code. This should never happens.

File: token/metadata.go (hunk 28 above)
Comment (KElkhiyaoui) [reply token/metadata.go:86]: This will be removed from here. A new issue will be opened to check well-formedness of token requests.

**Review by adecaro — APPROVED** [review]

**Pull request merged**